cel/validator.go

```diff
@@ -0,0 +1,282 @@
+	errs := errorReporter{iss: iss, info: a.info}
+	root := ast.NavigateCheckedAST(astToCheckedAST(a))
+	funcCalls := ast.MatchDescendants(root, ast.FunctionMatcher(v.funcName))
+	for _, call := range funcCalls {
+		callArgs := call.AsCall().Args()
+		if len(callArgs) <= v.argNum {
+			continue
+		}
+		litArg := callArgs[v.argNum]
+		if litArg.Kind() != ast.LiteralKind {
+			continue
+		}
+		if err := v.check(e, call, litArg); err != nil {
+			errs.reportErrorAtID(litArg.ID(), "invalid %s argument", v.funcName)
+		}
+	}
+}
+
+func evalCall(env *Env, call, arg ast.NavigableExpr) error {
+	ast := ParsedExprToAst(&exprpb.ParsedExpr{Expr: call.ToExpr()})
+	prg, err := env.Program(ast)
+	if err != nil {
+		return err
+	}
+	_, _, err = prg.Eval(NoVars())
+	return err
+}
+
+func compileRegex(_ *Env, _, arg ast.NavigableExpr) error {
+	pattern := arg.AsLiteral().Value().(string)
+	_, err := regexp.Compile(pattern)
+	return err
+}
+
+func (formatValidator) is_validator() {}
+
+type homogeneousAggregateLiteralValidator struct{}
+
+// Name returns the unique name of the homogeneous type validator.
+func (homogeneousAggregateLiteralValidator) Name() string {
+	return "cel.lib.std.types.homogeneous"
+}
+
+// Validate validates that all lists and map literals have homogeneous types, i.e. don't contain dyn types.
+//
+// This validator makes an exception for list and map literals which occur at any level of nesting within
+// string format calls.
+func (v homogeneousAggregateLiteralValidator) Validate(_ *Env, a *Ast, iss *Issues) {
+	errs := errorReporter{iss: iss, info: a.info}
+	root := ast.NavigateCheckedAST(astToCheckedAST(a))
+	listExprs := ast.MatchDescendants(root, ast.KindMatcher(ast.ListKind))
+	for _, listExpr := range listExprs {
+		if hasStringFormatAncestor(listExpr) {
+			continue
+		}
+		l := listExpr.AsList()
+		elements := l.Elements()
+		optIndices := l.OptionalIndices()
+		var elemType *Type
+		for i, e := range elements {
+			et := e.Type()
+			if isOptionalIndex(i, optIndices) {
+				et = et.Parameters()[0]
+			}
+			if elemType == nil {
+				elemType = et
+				continue
+			}
+			if !elemType.IsEquivalentType(et) {
+				v.typeMismatch(errs, e.ID(), elemType, et)
+				break
+			}
+		}
+	}
+	mapExprs := ast.MatchDescendants(root, ast.KindMatcher(ast.MapKind))
+	for _, mapExpr := range mapExprs {
+		if hasStringFormatAncestor(mapExpr) {
+			continue
+		}
+		m := mapExpr.AsMap()
+		entries := m.Entries()
+		var keyType, valType *Type
+		for _, e := range entries {
+			key, val := e.Key(), e.Value()
+			kt, vt := key.Type(), val.Type()
+			if e.IsOptional() {
+				vt = vt.Parameters()[0]
+			}
+			if keyType == nil && valType == nil {
+				keyType, valType = kt, vt
+				continue
+			}
+			if !keyType.IsEquivalentType(kt) {
+				v.typeMismatch(errs, key.ID(), keyType, kt)
+			}
+			if !valType.IsEquivalentType(vt) {
+				v.typeMismatch(errs, val.ID(), valType, vt)
+			}
+		}
+	}
+}
+
+func hasStringFormatAncestor(e ast.NavigableExpr) bool {
```

Comment:
I'm a bit torn on this. IMO, I think this type of check is too specific for homogeneous literal validation to be concerned of, especially considering that `string.format` is not part of the standard library. I think I would've preferred an approach where the string formatter validator composes homogenous literal validator then decide how to validate.

Reply:
Since I don't know when the validators will be applied (or if they're applied), I think I have to have a carve-out here for format checking. I can make it more dynamic in the sense that it can check for library enablement to determine what to do, but I'm not sure I can do much better than that. 

If you switch the order of `ext.Strings()` and `cel.HomogeneousAggregateLiterals()` in the `NewEnv()` call, I think there's a risk that the format validator might not see the aggregate literal validator or vice-versa. The only other thing I can do is allow for replacement of the validator, but that's a bit risky since the order in which the validators are added would influence the outcome.

Reply:
I'm okay with adding a TODO and coming back to this on a subsequent PR. I thought maybe I could make the check generic in the sense that if the specific overload expects a `list(dyn)` element type (I believe format is one of the few, possibly only), then homogeneous aggregate literal checks could be disabled since that was the developer's intent, but I'm not sure how string.format does it will be how other functions expect it to work.